meta-store/src/meta_store.rs

```diff
@@ -0,0 +1,193 @@
+/*
+    Quickwit
+    Copyright (C) 2021 Quickwit Inc.
+
+    Quickwit is offered under the AGPL v3.0 and as commercial software.
+    For commercial licensing, contact us at hello@quickwit.io.
+
+    AGPL:
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use std::collections::HashMap;
+use std::fmt::{Debug, Display};
+use std::io;
+use std::ops::Range;
+
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+// pub type IndexId = String;
+
+pub type SplitId = String;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct IndexMetaData {
+    version: String,
+    name: String,
+    path: String,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SplitMetaData {
+    // Split uri. In spirit, this uri should be self sufficient
+    // to identify a split.
+    // In reality, some information may be implicitly configure
+    // in the store uri resolver, such as the Amazon S3 region.
+    split_uri: String,
+
+    // Number of records (or document) in the split
+    num_records: u64,
+
+    // Weight in bytes of the split
+    size_in_bytes: u64,
+
+    // if a time field is available, the min / max timestamp in the segment.
+    time_range: Option<Range<u64>>,
+
+    // Number of merge this segment has been subjected to during its lifetime.
+    generation: usize,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct ManifestEntry {
+    file_name: String,
+    file_size_in_bytes: u64,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum SplitState {
+    // the splits is almost ready. Some of its files may have been uploaded in the storage.
+    Staged,
+    // the splits is ready and published.
+    Published,
+    // the split is scheduled to be deleted.
+    ScheduledForDeleted,
+}
+
+/// A split manifest carries all meta data about a split
+/// and its files.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SplitManifest {
```

Comment:
When holding a SplitManifest, i have no way of knowing the SplitId. I think we need to have the id here. Also do we want  to separate the slit_uri from the split_id. we could just strop the uri and have a method that extract the splt_id from the split_uri @fulmicoton @guilload